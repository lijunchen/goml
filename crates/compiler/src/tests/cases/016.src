enum List[T] {
    Nil,
    Cons(T, List[T]),
}

fn list_length[T](xs: List[T]) -> Int {
    match xs {
        Nil => 0,
        Cons(_, tail) => int_add(1, list_length(tail)),
    }
}

fn int_list_length(xs: List[Int]) -> Int {
    match xs {
        Nil => 0,
        Cons(_, tail) => int_add(1, int_list_length(tail)),
    }
}

fn main() {
    let x = Cons(1, Nil) in
    let length = list_length(x) in
    let _ = println(int_to_string(length)) in

    let x = Cons(1, Cons(2, Nil)) in
    let length = list_length(x) in
    let _ = println(int_to_string(length)) in

    let x = Cons(0, Cons(1, Cons(2, Nil))) in
    let length = int_list_length(x) in
    let _ = println(int_to_string(length)) in

    let x = Cons((), Nil) in
    let length = list_length(x) in
    let _ = println(int_to_string(length)) in

    let x = Cons((), Cons((), Nil)) in
    let length = list_length(x) in
    let _ = println(int_to_string(length)) in

    let x = Cons(true, Cons(false, Nil)) in
    let length = list_length(x) in
    let _ = println(int_to_string(length)) in
    ()
}
