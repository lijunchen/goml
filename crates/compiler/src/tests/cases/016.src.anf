fn list_length(xs/0: List[T]) -> Int {
  match xs/0 {
    Constr_0 => {
      0
    },
    Constr_1 => {
      let x0 = xs/0.constr.1.0 in
      let x1 = xs/0.constr.1.1 in
      let x0 = xs/0.constr.1.0 in
      let x1 = xs/0.constr.1.1 in
      let tail/1 = x1 in
      let t10 = list_length(tail/1) in
      int_add(1, t10)
    },
  }
}

fn int_list_length(xs/2: List[Int]) -> Int {
  match xs/2 {
    Constr_0 => {
      0
    },
    Constr_1 => {
      let x2 = xs/2.constr.1.0 in
      let x3 = xs/2.constr.1.1 in
      let x2 = xs/2.constr.1.0 in
      let x3 = xs/2.constr.1.1 in
      let tail/3 = x3 in
      let t11 = int_list_length(tail/3) in
      int_add(1, t11)
    },
  }
}

fn main() -> Unit {
  let t12 = Constr_0 in
  let x/4 = List::Cons(1, t12) in
  let length/5 = list_length(x/4) in
  let t13 = int_to_string(length/5) in
  let mtmp4 = println(t13) in
  let t15 = Constr_0 in
  let t14 = List::Cons(2, t15) in
  let x/6 = List::Cons(1, t14) in
  let length/7 = list_length(x/6) in
  let t16 = int_to_string(length/7) in
  let mtmp5 = println(t16) in
  let t19 = Constr_0 in
  let t18 = List::Cons(2, t19) in
  let t17 = List::Cons(1, t18) in
  let x/8 = List::Cons(0, t17) in
  let length/9 = int_list_length(x/8) in
  let t20 = int_to_string(length/9) in
  let mtmp6 = println(t20) in
  let t21 = Constr_0 in
  let x/10 = List::Cons((), t21) in
  let length/11 = list_length(x/10) in
  let t22 = int_to_string(length/11) in
  let mtmp7 = println(t22) in
  let t24 = Constr_0 in
  let t23 = List::Cons((), t24) in
  let x/12 = List::Cons((), t23) in
  let length/13 = list_length(x/12) in
  let t25 = int_to_string(length/13) in
  let mtmp8 = println(t25) in
  let t27 = Constr_0 in
  let t26 = List::Cons(false, t27) in
  let x/14 = List::Cons(true, t26) in
  let length/15 = list_length(x/14) in
  let t28 = int_to_string(length/15) in
  let mtmp9 = println(t28) in
  ()
}