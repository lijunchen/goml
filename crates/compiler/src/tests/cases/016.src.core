fn list_length(xs/0: List[T]) -> Int {
  match xs/0 {
    List::Nil => {
      0
    },
    List::Cons(x0, x1) => {
      let x0 = List_1_0(xs/0) in
      let x1 = List_1_1(xs/0) in
      let tail/1 = x1 in
      int_add(1, list_length(tail/1))
    },
  }
}

fn int_list_length(xs/2: List[Int]) -> Int {
  match xs/2 {
    List::Nil => {
      0
    },
    List::Cons(x2, x3) => {
      let x2 = List_1_0(xs/2) in
      let x3 = List_1_1(xs/2) in
      let tail/3 = x3 in
      int_add(1, int_list_length(tail/3))
    },
  }
}

fn main() -> Unit {
  let x/4 = List::Cons(1, List::Nil) in
  let length/5 = list_length(x/4) in
  let mtmp4 = string_println(int_to_string(length/5)) in
  let x/6 = List::Cons(1, List::Cons(2, List::Nil)) in
  let length/7 = list_length(x/6) in
  let mtmp5 = string_println(int_to_string(length/7)) in
  let x/8 = List::Cons(0, List::Cons(1, List::Cons(2, List::Nil))) in
  let length/9 = int_list_length(x/8) in
  let mtmp6 = string_println(int_to_string(length/9)) in
  let x/10 = List::Cons((), List::Nil) in
  let length/11 = list_length(x/10) in
  let mtmp7 = string_println(int_to_string(length/11)) in
  let x/12 = List::Cons((), List::Cons((), List::Nil)) in
  let length/13 = list_length(x/12) in
  let mtmp8 = string_println(int_to_string(length/13)) in
  let x/14 = List::Cons(true, List::Cons(false, List::Nil)) in
  let length/15 = list_length(x/14) in
  let mtmp9 = string_println(int_to_string(length/15)) in
  ()
}