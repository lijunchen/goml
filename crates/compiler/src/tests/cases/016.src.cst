FILE@0..98
  ENUM@0..49
    EnumKeyword@0..4 "enum"
    Whitespace@4..5 " "
    Uident@5..9 "List"
    GENERIC_LIST@9..13
      LBracket@9..10 "["
      GENERIC@10..11
        Uident@10..11 "T"
      RBracket@11..12 "]"
      Whitespace@12..13 " "
    LBrace@13..14 "{"
    Whitespace@14..19 "\n    "
    VARIANT_LIST@19..49
      VARIANT@19..22
        Uident@19..22 "Nil"
      Comma@22..23 ","
      Whitespace@23..28 "\n    "
      VARIANT@28..44
        Uident@28..32 "Cons"
        TYPE_LIST@32..44
          LParen@32..33 "("
          TYPE_TAPP@33..34
            Uident@33..34 "T"
          Comma@34..35 ","
          Whitespace@35..36 " "
          TYPE_TAPP@36..43
            Uident@36..40 "List"
            GENERIC_LIST@40..43
              LBracket@40..41 "["
              GENERIC@41..42
                Uident@41..42 "T"
              RBracket@42..43 "]"
          RParen@43..44 ")"
      Comma@44..45 ","
      Whitespace@45..46 "\n"
      RBrace@46..47 "}"
      Whitespace@47..49 "\n\n"
  FN@49..98
    FnKeyword@49..51 "fn"
    Whitespace@51..52 " "
    Lident@52..56 "main"
    PARAM_LIST@56..59
      LParen@56..57 "("
      RParen@57..58 ")"
      Whitespace@58..59 " "
    BLOCK@59..98
      LBrace@59..60 "{"
      Whitespace@60..65 "\n    "
      EXPR_LET@65..96
        LetKeyword@65..68 "let"
        Whitespace@68..69 " "
        PATTERN_VARIABLE@69..71
          Lident@69..70 "x"
          Whitespace@70..71 " "
        Eq@71..72 "="
        Whitespace@72..73 " "
        EXPR_LET_VALUE@73..86
          EXPR_CALL@73..86
            EXPR_UIDENT@73..77
              Uident@73..77 "Cons"
            ARG_LIST@77..86
              LParen@77..78 "("
              ARG@78..81
                EXPR_INT@78..79
                  Int32@78..79 "1"
                Comma@79..80 ","
                Whitespace@80..81 " "
              ARG@81..84
                EXPR_UIDENT@81..84
                  Uident@81..84 "Nil"
              RParen@84..85 ")"
              Whitespace@85..86 " "
        InKeyword@86..88 "in"
        Whitespace@88..93 "\n    "
        EXPR_LET_BODY@93..96
          EXPR_UNIT@93..96
            LParen@93..94 "("
            RParen@94..95 ")"
            Whitespace@95..96 "\n"
      RBrace@96..97 "}"
      Whitespace@97..98 "\n"