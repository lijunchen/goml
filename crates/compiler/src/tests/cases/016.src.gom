package main

import (
    "fmt"
)

func unit_to_string(x struct{}) string {
    return "()"
}

func bool_to_string(x bool) string {
    if x {
        return "true"
    } else {
        return "false"
    }
}

func int_to_string(x int) string {
    return fmt.Sprintf("%d", x)
}

func int_add(x int, y int) int {
    return x + y
}

func int_sub(x int, y int) int {
    return x - y
}

func int_less(x int, y int) bool {
    return x < y
}

func string_print(s string) struct{} {
    fmt.Print(s)
    return struct{}{}
}

func string_println(s string) struct{} {
    fmt.Println(s)
    return struct{}{}
}

func missing(s string) struct{} {
    panic("missing: " + s)
    return struct{}{}
}

type List__Int interface {
    isList__Int()
}

type Nil struct {}

func (_ Nil) isList__Int() {}

type Cons struct {
    _0 int
    _1 List__Int
}

func (_ Cons) isList__Int() {}

type List__TParam_T interface {
    isList__TParam_T()
}

type Nil struct {}

func (_ Nil) isList__TParam_T() {}

type Cons struct {
    _0 T
    _1 List__TParam_T
}

func (_ Cons) isList__TParam_T() {}

type List__Unit interface {
    isList__Unit()
}

type Nil struct {}

func (_ Nil) isList__Unit() {}

type Cons struct {
    _0 struct{}
    _1 List__Unit
}

func (_ Cons) isList__Unit() {}

type List__Bool interface {
    isList__Bool()
}

type Nil struct {}

func (_ Nil) isList__Bool() {}

type Cons struct {
    _0 bool
    _1 List__Bool
}

func (_ Cons) isList__Bool() {}

func int_list_length(xs__2 List__Int) int {
    var ret31 int
    switch xs__2 := xs__2.(type) {
    case Nil:
        ret31 = 0
    case Cons:
        var x3 List__TParam_T = xs__2._1
        var tail__3 List__Int = x3
        var t10 int = int_list_length(tail__3)
        ret31 = int_add(1, t10)
    }
    return ret31
}

func main0() struct{} {
    var ret32 struct{}
    var t11 List__Int = Nil{}
    var x__4 List__Int = Cons{
        _0: 1,
        _1: t11,
    }
    var length__5 int = list_length__T_Int(x__4)
    var t12 string = int_to_string(length__5)
    string_println(t12)
    var t14 List__Int = Nil{}
    var t13 List__Int = Cons{
        _0: 2,
        _1: t14,
    }
    var x__6 List__Int = Cons{
        _0: 1,
        _1: t13,
    }
    var length__7 int = list_length__T_Int(x__6)
    var t15 string = int_to_string(length__7)
    string_println(t15)
    var t18 List__Int = Nil{}
    var t17 List__Int = Cons{
        _0: 2,
        _1: t18,
    }
    var t16 List__Int = Cons{
        _0: 1,
        _1: t17,
    }
    var x__8 List__Int = Cons{
        _0: 0,
        _1: t16,
    }
    var length__9 int = int_list_length(x__8)
    var t19 string = int_to_string(length__9)
    string_println(t19)
    var t20 List__Unit = Nil{}
    var x__10 List__Unit = Cons{
        _0: struct{}{},
        _1: t20,
    }
    var length__11 int = list_length__T_Unit(x__10)
    var t21 string = int_to_string(length__11)
    string_println(t21)
    var t23 List__Unit = Nil{}
    var t22 List__Unit = Cons{
        _0: struct{}{},
        _1: t23,
    }
    var x__12 List__Unit = Cons{
        _0: struct{}{},
        _1: t22,
    }
    var length__13 int = list_length__T_Unit(x__12)
    var t24 string = int_to_string(length__13)
    string_println(t24)
    var t26 List__Bool = Nil{}
    var t25 List__Bool = Cons{
        _0: false,
        _1: t26,
    }
    var x__14 List__Bool = Cons{
        _0: true,
        _1: t25,
    }
    var length__15 int = list_length__T_Bool(x__14)
    var t27 string = int_to_string(length__15)
    string_println(t27)
    ret32 = struct{}{}
    return ret32
}

func list_length__T_Int(xs__0 List__Int) int {
    var ret33 int
    switch xs__0 := xs__0.(type) {
    case Nil:
        ret33 = 0
    case Cons:
        var x1 List__Int = xs__0._1
        var tail__1 List__Int = x1
        var t28 int = list_length__T_Int(tail__1)
        ret33 = int_add(1, t28)
    }
    return ret33
}

func list_length__T_Unit(xs__0 List__Unit) int {
    var ret34 int
    switch xs__0 := xs__0.(type) {
    case Nil:
        ret34 = 0
    case Cons:
        var x1 List__Unit = xs__0._1
        var tail__1 List__Unit = x1
        var t29 int = list_length__T_Unit(tail__1)
        ret34 = int_add(1, t29)
    }
    return ret34
}

func list_length__T_Bool(xs__0 List__Bool) int {
    var ret35 int
    switch xs__0 := xs__0.(type) {
    case Nil:
        ret35 = 0
    case Cons:
        var x1 List__Bool = xs__0._1
        var tail__1 List__Bool = x1
        var t30 int = list_length__T_Bool(tail__1)
        ret35 = int_add(1, t30)
    }
    return ret35
}

func main() {
    main0()
}
