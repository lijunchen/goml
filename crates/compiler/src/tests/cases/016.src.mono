fn int_list_length(xs/2: List__Int) -> Int {
  match xs/2 {
    List__Int::Nil => {
      0
    },
    List__Int::Cons(x2, x3) => {
      let x2 = List__Int_1_0(xs/2) in
      let x3 = List__Int_1_1(xs/2) in
      let tail/3 = x3 in
      int_add(1, int_list_length(tail/3))
    },
  }
}

fn main() -> Unit {
  let x/4 = List__Int::Cons(1, List__Int::Nil) in
  let length/5 = list_length__T_Int(x/4) in
  let mtmp4 = string_println(int_to_string(length/5)) in
  let x/6 = List__Int::Cons(1, List__Int::Cons(2, List__Int::Nil)) in
  let length/7 = list_length__T_Int(x/6) in
  let mtmp5 = string_println(int_to_string(length/7)) in
  let x/8 = List__Int::Cons(0, List__Int::Cons(1, List__Int::Cons(2, List__Int::Nil))) in
  let length/9 = int_list_length(x/8) in
  let mtmp6 = string_println(int_to_string(length/9)) in
  let x/10 = List__Unit::Cons((), List__Unit::Nil) in
  let length/11 = list_length__T_Unit(x/10) in
  let mtmp7 = string_println(int_to_string(length/11)) in
  let x/12 = List__Unit::Cons((), List__Unit::Cons((), List__Unit::Nil)) in
  let length/13 = list_length__T_Unit(x/12) in
  let mtmp8 = string_println(int_to_string(length/13)) in
  let x/14 = List__Bool::Cons(true, List__Bool::Cons(false, List__Bool::Nil)) in
  let length/15 = list_length__T_Bool(x/14) in
  let mtmp9 = string_println(int_to_string(length/15)) in
  ()
}

fn list_length__T_Int(xs/0: List__Int) -> Int {
  match xs/0 {
    List__Int::Nil => {
      0
    },
    List__Int::Cons(x0, x1) => {
      let x0 = List__Int_1_0(xs/0) in
      let x1 = List__Int_1_1(xs/0) in
      let tail/1 = x1 in
      int_add(1, list_length__T_Int(tail/1))
    },
  }
}

fn list_length__T_Unit(xs/0: List__Unit) -> Int {
  match xs/0 {
    List__Unit::Nil => {
      0
    },
    List__Unit::Cons(x0, x1) => {
      let x0 = List__Unit_1_0(xs/0) in
      let x1 = List__Unit_1_1(xs/0) in
      let tail/1 = x1 in
      int_add(1, list_length__T_Unit(tail/1))
    },
  }
}

fn list_length__T_Bool(xs/0: List__Bool) -> Int {
  match xs/0 {
    List__Bool::Nil => {
      0
    },
    List__Bool::Cons(x0, x1) => {
      let x0 = List__Bool_1_0(xs/0) in
      let x1 = List__Bool_1_1(xs/0) in
      let tail/1 = x1 in
      int_add(1, list_length__T_Bool(tail/1))
    },
  }
}