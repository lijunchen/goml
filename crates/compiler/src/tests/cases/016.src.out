()
====
1
2
3
1
2
2

====
# command-line-arguments
./main.go:69:6: Nil redeclared in this block
	./main.go:54:6: other declaration of Nil
./main.go:73:6: Cons redeclared in this block
	./main.go:58:6: other declaration of Cons
./main.go:74:8: undefined: T
./main.go:84:6: Nil redeclared in this block
	./main.go:54:6: other declaration of Nil
./main.go:88:6: Cons redeclared in this block
	./main.go:58:6: other declaration of Cons
./main.go:99:6: Nil redeclared in this block
	./main.go:54:6: other declaration of Nil
./main.go:103:6: Cons redeclared in this block
	./main.go:58:6: other declaration of Cons
./main.go:116:33: cannot use xs__2._1 (variable of interface type List__Int) as List__TParam_T value in variable declaration: List__Int does not implement List__TParam_T (missing method isList__TParam_T)
./main.go:117:33: cannot use x3 (variable of interface type List__TParam_T) as List__Int value in variable declaration: List__TParam_T does not implement List__Int (missing method isList__Int)
./main.go:164:13: cannot use struct{}{} (value of type struct{}) as int value in struct literal
./main.go:164:13: too many errors
